onnxscript/values.py

```diff
@@ -271,12 +299,93 @@ def __init__(
         self.source = source
         self.kwargs = kwargs
         self._param_schemas: Optional[tuple[ParamSchema, ...]] = None
+        self._opschema: Optional[onnx.defs.OpSchema] = None
 
     @property
     def name(self):
         """Returns the function name."""
         return self.opname
 
+    @property
+    def opschema(self) -> Optional[onnx.defs.OpSchema]:
+        """Construct an OpSchema from function_ir."""
+        if self._opschema is not None:
+            return self._opschema
+
+        if not _ONNX_OP_SCHEMA_WRITABLE:
+            return None
+
+        function_ir = self.function_ir
+        # Find all distinct types in the inputs and outputs
```

Comment:
What's the notion of equality (of types) here? We need to be careful about the following: two parameters of type which is the same TypeVar can be mapped to the same typevar in ONNX. However, Python is more expressive. It allows us to define two parameters of type "Optional[TypeVar[X]]", right? ONNX doesn't allow us to express something like this. This is a partial constraint since, it allows one parameter to be None and the other to be non-None. But if both are non-None, they are expected to be of the same type. We can't say that in ONNX.

Reply:
Good catch 🤔 Do you have any suggestions? Should we be more relaxed about the optional inputs? Or maybe fix it when we need those scenarios? For now we mostly just handle static optionals

Reply:
I will add a comment that points to an issue for refinement in a separate PR.

Reply:
Ok. I think splitting the logic into the loop here and the utility `get_type_constraint_name` is problematic. One approach would be to combine/integrate both: that is, define a single function that maps a collection of types into a corresponding collection of type-var-name, plus bindings for those type-vars (if needed). Then, it would be able to map use of same python type-variable to the same onnx type-var, while renaming them into unique instances when used as a parameter to some other type-constructor like optional/list.

Reply:
Or, may be the renaming can be done even with the current split (may be integrating them is not necessary).

Reply:
I am not following 100%. Could you explain why it's problematic? Or I can chat with you offline

Reply:
It is problematic for the renaming part, since we would need some state-like mechanism to generate distinct names over several different inputs/outputs. But, on reflection, renaming won't be a perfect solution either. We can't capture the python type-constraint exactly in ONNX (in the problematic case). I was suggesting relaxing the type-constraint instead of tightening the type-constraint. But perhaps the preferable solution is to raise an error and let the use know about the issue.

Reply:
Currently, the logic is

- TypeVars -> use the name because it is unique for the function.
- Other types -> always create a new name, as intended in the signature
- Optional[TypeVar] -> Trickier: If input is Optional[TypeVar] and output is TypeVar, they are assigned to different type-names but should _sometimes_ be the same like you suggested. If we drop support for dynamic optionals, then we can assign the same name for `Optional[TypeVar]` and `TypeVar`, which is accurate. Nevertheless, assigning a different name for `Optional[TypeVar]` and `TypeVar` is a more relaxed constraint.

Reply:
I created #696. I am planning to merge this PR and refine based on what falls out from the discussion in the future.